subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/model/internal/IdeaDependenciesProvider.java

```diff
@@ -78,34 +73,128 @@ public IdeaDependenciesProvider(ServiceRegistry serviceRegistry) {
 
     private Set<Dependency> getDependencies(IdeaModule ideaModule) {
         Set<Dependency> dependencies = Sets.newLinkedHashSet();
+        Set<ModuleDependency> moduleDependencies = Sets.newLinkedHashSet();
+        Set<SingleEntryModuleLibrary> fileDependencies = Sets.newLinkedHashSet();
+
+
         for (GeneratedIdeaScope scope : GeneratedIdeaScope.values()) {
-            dependencies.addAll(getProjectDependencies(ideaModule, scope));
-            dependencies.addAll(getExternalDependencies(ideaModule, scope));
-            dependencies.addAll(getFileDependencies(ideaModule, scope));
+            moduleDependencies.addAll(getProjectDependencies(ideaModule, scope));
+            fileDependencies.addAll(getExternalDependencies(ideaModule, scope));
+            fileDependencies.addAll(getFileDependencies(ideaModule, scope));
         }
+
+        dependencies.addAll(optimizeDeps(moduleDependencies));
+        dependencies.addAll(optimizeDeps(fileDependencies, ideaModule));
         return dependencies;
     }
 
-    private Set<Dependency> getProjectDependencies(IdeaModule ideaModule, GeneratedIdeaScope scope) {
+    private Collection<Dependency> optimizeDeps(Collection<ModuleDependency> deps) {
+        Set<Dependency> result = Sets.newLinkedHashSet();
+        Multimap<String, GeneratedIdeaScope> namesToScopes = MultimapBuilder.hashKeys().linkedHashSetValues().build();
+        for (ModuleDependency dep : deps) {
+            namesToScopes.put(dep.getName(), GeneratedIdeaScope.valueOf(dep.getScope()));
+        }
+
+        for (ModuleDependency dep : deps) {
+            String moduleName = dep.getName();
+            Collection<GeneratedIdeaScope> ideaScopes = namesToScopes.get(moduleName);
```

Comment:
❌ There's lots of duplicated code between this method and the next.

Reply:
sorry for that. I've added the patch to deduplicate the code.